scipy/sparse/_coo.py

```diff
@@ -346,23 +337,32 @@ def tocsr(self, copy=False):
         if self.nnz == 0:
             return self._csr_container(self.shape, dtype=self.dtype)
         else:
-            M,N = self.shape
-            idx_dtype = self._get_index_dtype(self.coords, maxval=max(self.nnz, N))
-            row = self.row.astype(idx_dtype, copy=False)
-            col = self.col.astype(idx_dtype, copy=False)
-
-            indptr = np.empty(M + 1, dtype=idx_dtype)
-            indices = np.empty_like(col, dtype=idx_dtype)
-            data = np.empty_like(self.data, dtype=upcast(self.dtype))
-
-            coo_tocsr(M, N, self.nnz, row, col, self.data,
-                      indptr, indices, data)
+            from ._csr import csr_array
+            indptr, indices, data, shape = self._coo_to_compressed(csr_array._swap)
 
             x = self._csr_container((data, indices, indptr), shape=self.shape)
             if not self.has_canonical_format:
                 x.sum_duplicates()
             return x
 
+    def _coo_to_compressed(self, swap):
+        """convert (shape, coords, data) to (indptr, indices, data, shape)"""
+        M, N = swap(self.shape)
+        major, minor = swap(self.coords)
+        nnz = len(major)
+        # convert idx_dtype intc to int32 for pythran.
+        # tested in scipy/optimize/tests/test__numdiff.py::test_group_columns
+        idx_dtype = self._get_index_dtype(self.coords, maxval=max(self.nnz, M, N))
```

Comment:
Technically this should `max(self.nnz, N)`, to match the previous behavior.

Reply:
I think the previous behavior is possibly incorrect -- if M is e.g. bigger than 2**32 don't we want `idx_dtype` to be int64?  But I guess I should make that change in a different PR -- and maybe add a test too.

Reply:
Done!  The M is removed to match previous behavior.